lib/luaxtable.c

```diff
@@ -0,0 +1,234 @@
+* included in all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/module.h>
+#include <linux/version.h>
+#include <linux/types.h>
+#include <linux/netfilter.h>
+#include <linux/netfilter/x_tables.h>
+
+#include <lua.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+typedef enum luaxtable_type_e {
+	LUAXTABLE_TMATCH,
+	LUAXTABLE_TTARGET,
+} luaxtable_type_t;
+
+typedef struct luaxtable_s {
+	lunatik_object_t *runtime;
+	union {
+		struct xt_match *match;
+		struct xt_target *target;
+	};
+	luaxtable_type_t type;
+} luaxtable_t;
+
+static void luaxtable_release(void *private);
+
+#define luaxtable_setinteger(L,hook, field) 			\
+do {								\
+	lunatik_checkfield(L, 1, #field, LUA_TNUMBER);		\
+	hook->field = lua_tointeger(L, -1);			\
+} while (0)
+
+static inline const char * luaxtable_checkname(lua_State *L, int idx)
+{
+        const char *name;
+        size_t len;
+
+        lunatik_checkfield(L, idx, "name", LUA_TSTRING);
+        name = lua_tolstring(L, -1, &len);
+        if (len >= XT_EXTENSION_MAXNAMELEN)
+                luaL_error(L, "name is too long");
+        return name;
+}
+
+static inline luaxtable_t *luaxtable_new(lua_State *L, const lunatik_class_t *cls, int hook)
+{
+	lunatik_object_t *object = lunatik_newobject(L, cls, sizeof(luaxtable_t));
+	luaxtable_t *xtable = (luaxtable_t *)object->private;
+
+	xtable->runtime = lunatik_toruntime(L);
+	xtable->type = hook;
+	lunatik_getobject(xtable->runtime);
+	lunatik_registerobject(L, 1, object);
+	return xtable;
+}
+
+#define LUAXTABLE_NEWHOOK(hook, HOOK)					\
+static int luaxtable_new##hook(lua_State *L) 				\
+{									\
+	luaL_checktype(L, 1, LUA_TTABLE);				\
+									\
+	struct xt_##hook *hook = lunatik_checkalloc(L, sizeof(struct xt_##hook));	\
+	hook->me = THIS_MODULE;						\
+									\
+	const char *name = luaxtable_checkname(L, 1);			\
+	strncpy(hook->name, name, XT_EXTENSION_MAXNAMELEN);		\
+									\
+	luaxtable_setinteger(L, hook, revision);			\
+	luaxtable_setinteger(L, hook, family);				\
+	luaxtable_setinteger(L, hook, proto);				\
+									\
+	hook->hook = luaxtable_##hook;					\
+	hook->checkentry = luaxtable_##hook##_check;			\
+	hook->destroy = luaxtable_##hook##_destroy;			\
+									\
+	if (xt_register_##hook(hook) != 0) {				\
+		luaL_error(L, "unable to register" #hook);		\
+	}								\
+									\
+	luaxtable_t *xtable = luaxtable_new(L, &luaxtable_class, HOOK); \
+	xtable->hook = hook;						\
```

Comment:
for having "release()" being called in case of errors, you would need to call  "lunatik_newobject()" before calling "luaL_error()", because it's the one that pushes the object with the associated "release()" callback in the Lua stack.. perhaps you can move it to the beginning of this macro..
```
	luaL_checktype(L, 1, LUA_TTABLE);				\
									\
	lunatik_object_t *object = lunatik_newobject(L, &luaxtable_class, sizeof(luaxtable_t));	\
	/* errors after this line will eventually trigger "object->release()" */
	luaxtable_t *xtable = (luaxtable_t *)object->private;				\
									\
	struct xt_##hook *hook = lunatik_checkalloc(L, sizeof(struct xt_##hook));	\
```

you would probably need to check if it's safe to unregister a not registered hook.. if it's not, you need have some guard for that on your release callback.. because this function might fail before or after registering..